edgelet/edgelet-http-workload/src/server/decrypt.rs

```diff
@@ -48,255 +53,57 @@ where
                 req.into_body().concat2().then(|body| {
                     let body =
                         body.context(ErrorKind::EncryptionOperation(EncryptionOperation::Decrypt))?;
-                    Ok((id, body))
+                    let request: DecryptRequest =
+                        serde_json::from_slice(&body).context(ErrorKind::MalformedRequestBody)?;
+                    Ok((id, request))
                 })
             })
             .into_future()
             .flatten()
-            .and_then(move |(id, body)| -> Result<_, Error> {
-                let request: DecryptRequest =
-                    serde_json::from_slice(&body).context(ErrorKind::MalformedRequestBody)?;
-                let ciphertext = base64::decode(request.ciphertext())
-                    .context(ErrorKind::MalformedRequestBody)?;
+            .and_then(move |(id, request)| -> Result<_, Error> {
+                let plaintext =
+                    base64::decode(request.ciphertext()).context(ErrorKind::MalformedRequestBody)?;
                 let initialization_vector = base64::decode(request.initialization_vector())
                     .context(ErrorKind::MalformedRequestBody)?;
-                let plaintext = hsm
-                    .decrypt(id.as_bytes(), &ciphertext, &initialization_vector)
-                    .context(ErrorKind::EncryptionOperation(EncryptionOperation::Decrypt))?;
-                let encoded = base64::encode(&plaintext);
-                let response = DecryptResponse::new(encoded);
-                let body = serde_json::to_string(&response)
-                    .context(ErrorKind::EncryptionOperation(EncryptionOperation::Decrypt))?;
-                let response = Response::builder()
-                    .status(StatusCode::OK)
-                    .header(CONTENT_TYPE, "application/json")
-                    .header(CONTENT_LENGTH, body.len().to_string().as_str())
-                    .body(body.into())
-                    .context(ErrorKind::EncryptionOperation(EncryptionOperation::Decrypt))?;
-                Ok(response)
+                let ciphertext = get_key_handle(id_mgr, &id)
```

Comment:
Currently the module encryption key is not the same as the module ID key. Just like module ID key is derived from master ID key (device ID key now) + module metadata, the module encryption key is derived from the master encryption key + module metadata.

Is it intentional that you're making them both the same?

Reply:
Fixed it so we create a new master encryption key, if it doesn't exist